File: src/entities/Collision.java
package entities;

import java.util.ArrayList;
import java.util.List;

import org.joml.Vector2d;

import entities.Hitbox;
import map.Map;

public class Collision{
	public static final int X_COLLISION = 0b01;
	public static final int Y_COLLISION = 0b10;
	
	public List<Vector2d> collisionPos;
	public Vector2d entityPos;
	public double collisionStrength;
	public int collisionFlags;
	
	public Collision(List<Vector2d> collisionPos, Vector2d entityPos, double collisionStrength, int collisionFlags) {
		this.collisionPos = collisionPos;
		this.entityPos = entityPos;
		this.collisionStrength = collisionStrength;
		this.collisionFlags = collisionFlags;
	}
	
	public static Collision canMoveTo(Map map, Hitbox hitbox, double originX, double originY, Vector2d m){
		return canMoveTo(map, hitbox, originX, originY, m, false);
	}
	
	private static final double stepsize = 0.5;
	
	public static Collision canMoveTo(Map map, Hitbox hitbox, double originX, double originY, Vector2d m, boolean canSlip){
		int interaction = getInteraction(m);
		int collisionFlags = 0;
		List<Vector2d> collisionPos = new ArrayList<>();
		double l = m.length();
		double newX = originX, newY = originY, lastX = originX, lastY = originY;
//		System.out.printf("eX:%f eY:%f sX:%f sY:%f\n", c.entityPos.x, c.entityPos.y, speedX, speedY);
		MainCheck:
		for (double sl = 0; sl <= l+stepsize+0.1; sl+=stepsize) {
			if(sl<=l) {
				newX = originX + sl/l*m.x;
				newY = originY + sl/l*m.y;
			}else {
				newX = originX + m.x;
				newY = originY + m.y;
			}
//			System.out.printf("eX:%f eY:%f sX:%f sY:%f\n", c.entityPos.x, c.entityPos.y, speedX, speedY);
			//gets the first Collision on the Y axis
			if((collisionFlags & Collision.Y_COLLISION) == 0 && m.y!=0){
				for(double i = hitbox.xmin; i <= hitbox.xmax; i+=0.5){
					if(i>hitbox.xmax)i = hitbox.xmax;
					if((map.getSolidity(lastX+i,newY+(m.y<0 ? hitbox.ymin : hitbox.ymax)) & interaction) != 0) {
/*						if(canSlip) {
							if((collisionFlags & Collision.X_COLLISION) == 0 && getCollision(map, hitbox, interaction, null, lastX+(m.x < 0 ? -1 : 1), newY, lastX, lastY) == 0) {
								lastX = lastX + (m.x < 0 ? -1 : 1);
								originX += (m.x < 0 ? -1 : 1) - stepsize/l*m.x;
								lastY = newY;
								continue MainCheck;
							}else if(getCollision(map, hitbox, interaction ^ Map.SOLID_X_AXIS, null, lastX+(m.x < 0 ? 1 : -1), newY, lastX, lastY) == 0){
								lastX = lastX + (m.x < 0 ? 1 : -1);
								originX += (m.x < 0 ? 1 : -1) - stepsize/l*m.x;
								lastY = newY;
								continue MainCheck;
							}
						}*/
						collisionPos.add(new Vector2d(lastX+i,newY+(m.y<0 ? hitbox.ymin : hitbox.ymax)));
						collisionFlags |= Y_COLLISION;
					}
//					System.out.printf("pX:%f pY:%f ccc\n", x+i,y+(m.y<0 ? hitbox.ymin : hitbox.ymax));
				}
			}
			
			//gets the first Collision on the X axis
			if((collisionFlags & Collision.X_COLLISION) == 0 && m.x!=0){
				for(double i = hitbox.ymin; i <= hitbox.ymax; i+=0.5){
					if(i>hitbox.ymax)i = hitbox.ymax;
					if((map.getSolidity(newX+(m.x<0 ? hitbox.xmin : hitbox.xmax),lastY+i) & interaction) != 0) {
						if(canSlip) {
							if((collisionFlags & Collision.Y_COLLISION) == 0 && getCollision(map, hitbox, interaction, null, newX, lastY+(m.y < 0 ? -1 : 1), lastX, lastY) == 0) {
								lastY = lastY + (m.y < 0 ? -1 : 1);
								originY += (m.y < 0 ? -1 : 1) - stepsize/l*m.y;
								lastX = newX;
								continue MainCheck;
							}else if(getCollision(map, hitbox, interaction ^ Map.SOLID_Y_AXIS, null, newX, lastY+(m.y < 0 ? 1 : -1), lastX, lastY) == 0){
								lastY = lastY + (m.y < 0 ? 1 : -1);
								originY += (m.y < 0 ? 1 : -1) - stepsize/l*m.y;
								lastX = newX;
								continue MainCheck;
							}
						}
						collisionPos.add(new Vector2d(newX+(m.x<0 ? hitbox.xmin : hitbox.xmax),lastY+i));
						collisionFlags |= X_COLLISION;
					}
//					System.out.printf("pX:%f pY:%f ccc\n", x+(m.x<0 ? hitbox.xmin : hitbox.xmax),y+i);
				}
			}
  			if((collisionFlags & Collision.X_COLLISION) > 0 && (collisionFlags & Collision.Y_COLLISION) > 0) {
  				break;
			}else {
				if((collisionFlags & Collision.X_COLLISION) == 0)lastX = newX;
				if((collisionFlags & Collision.Y_COLLISION) == 0)lastY = newY;
			}
		}
		if(collisionPos.size()>0 || canSlip) {
			return new Collision(collisionPos, new Vector2d(lastX, lastY), l, collisionFlags);
		}else {
			return null;
		}
	}
	
	public static int getCollision(Map map, Hitbox hitbox, int interaction, List<Vector2d> collisionPos, double newX, double newY, double lastX, double lastY) {
		int collisionFlags = 0;
		for(double i = hitbox.xmin; i <= hitbox.xmax; i+=0.5){
			if(i>hitbox.xmax)i = hitbox.xmax;
			if((map.getSolidity(newX+i,newY+(newY < lastY ? hitbox.ymin : hitbox.ymax)) & interaction) != 0) {
				if(collisionPos != null)collisionPos.add(new Vector2d(newX+i,newY+(newY < lastY ? hitbox.ymin : hitbox.ymax)));
				collisionFlags |= Y_COLLISION;
			}
		}
		
		for(double i = hitbox.ymin; i <= hitbox.ymax; i+=0.5){
			if(i>hitbox.ymax)i = hitbox.ymax;
			if((map.getSolidity(newX+(newX < lastX ? hitbox.xmin : hitbox.xmax),newY+i) & interaction) != 0) {
				if(collisionPos != null)collisionPos.add(new Vector2d(newX+(newX < lastX ? hitbox.xmin : hitbox.xmax),newY+i));
				collisionFlags |= X_COLLISION;
			}
		}
		
		return collisionFlags;
	}
	
	public static int getInteraction(Vector2d vec) {
		return (vec.x<0 ? Map.SOLID_RIGHT : vec.x>0 ? Map.SOLID_LEFT : 0) | (vec.y<0 ? Map.SOLID_BOTTOM : vec.y>0 ? Map.SOLID_TOP : 0);
	}
}

File: project/src/entities/Collision.cpp
#include "Collision.h"

#include "../main/Game.h"

namespace Pixelverse {

Collision::Collision() : collide_x(false), collide_y(false), collisionStrength(0){}

Collision::~Collision(){}

std::unique_ptr<Collision> Collision::canMoveTo(Hitbox hitbox, vec2 position, vec2 movement, double rotation = 0, bool canSlip = false){
	std::unique_ptr<Collision> collision = std::make_unique<Collision>();

	double l = movement.length();
	rotation = fmod(rotation, M_PI*2);
	bool sideways = (rotation > M_PI/4 && rotation < M_PI/4*3) || (rotation < -M_PI/4 && rotation > -M_PI/4*3);
	if(sideways) hitbox = {hitbox.ymin, hitbox.ymax, hitbox.xmin, hitbox.xmax};
	vec2 newPos(position), lastPos(position);
	MainCheck:
	for (double sl = 0; sl <= l + stepsize + 0.1; sl += stepsize) {
		if(sl<=l) {
			newPos = position + movement * sl/l;
		}else {
			newPos = position + movement;
		}
		//gets the first Collision on the Y axis
		if(!collision->collide_y && movement.y != 0.0){
			for(double i = hitbox.xmin; i <= hitbox.xmax; i+=0.5){
				if(i>hitbox.xmax)i = hitbox.xmax;
				coordinate check = coordinate(vec2{lastPos.x+i,newPos.y+(movement.y<0 ? hitbox.ymin : hitbox.ymax)});
				if(Game::map->getMaterial(check, true)->solid) {
					if(canSlip && sideways) {
						if(!collision->collide_x && getCollision(hitbox, vec2{newPos.x + (movement.x < 0 ? -2 : 2), lastPos.y}, lastPos)) {
							lastPos.x = lastPos.x + (movement.x < 0 ? -2 : 2);
							position.x += (movement.x < 0 ? -2 : 2) - stepsize/l*movement.x;
							lastPos.y = newPos.y;
							goto MainCheck;
						}else if(getCollision(hitbox, vec2{newPos.x + (movement.x < 0 ? 2 : -2), lastPos.y}, lastPos)){
							lastPos.x = lastPos.x + (movement.x < 0 ? 2 : -2);
							position.x += (movement.x < 0 ? 2 : -2) - stepsize/l*movement.x;
							lastPos.y = newPos.y;
							goto MainCheck;
						}
					}
					collision->collisions.push_back(check);
					collision->collide_y = true;
				}
			}
		}

		//gets the first Collision on the X axis
		if(!collision->collide_x && movement.x != 0.0){
			for(double i = hitbox.ymin; i <= hitbox.ymax; i+=0.5){
				if(i>hitbox.ymax)i = hitbox.ymax;
				coordinate check = coordinate(vec2{newPos.x+(movement.x<0 ? hitbox.xmin : hitbox.xmax),lastPos.y+i});
				if(Game::map->getMaterial(check, true)->solid) {
					if(canSlip && !sideways) {
						if(!collision->collide_y && getCollision(hitbox, vec2{newPos.x, lastPos.y + (movement.y < 0 ? -2 : 2)}, lastPos)) {
							lastPos.y = lastPos.y + (movement.y < 0 ? -2 : 2);
							position.y += (movement.y < 0 ? -2 : 2) - stepsize/l*movement.y;
							lastPos.x = newPos.x;
							goto MainCheck;
						}else if(getCollision(hitbox, vec2{newPos.x, lastPos.y+(movement.y < 0 ? 2 : -2)}, lastPos)){
							lastPos.y = lastPos.y + (movement.y < 0 ? 2 : -2);
							position.y += (movement.y < 0 ? 2 : -2) - stepsize/l*movement.y;
							lastPos.x = newPos.x;
							goto MainCheck;
						}
					}
					collision->collisions.push_back(check);
					collision->collide_x = true;
				}
			}
		}
		if(collision->collide_x && collision->collide_y) {
			break;
		}else {
			if(!collision->collide_x)lastPos.x = newPos.x;
			if(!collision->collide_y)lastPos.y = newPos.y;
		}
	}
	collision->entityPos = lastPos;
	collision->collisionStrength = l;

	return collision;
}

bool Collision::getCollision(Hitbox hitbox, vec2 newPos, vec2 lastPos) {
	for(double i = hitbox.xmin; i <= hitbox.xmax; i+=0.5){
		if(i>hitbox.xmax)i = hitbox.xmax;
		if(Game::map->getMaterial(coordinate(vec2{newPos.x+i,newPos.y+(newPos.y < lastPos.y ? hitbox.ymin : hitbox.ymax)}), true)->solid) {
			return false;
		}
	}

	for(double i = hitbox.ymin; i <= hitbox.ymax; i+=0.5){
		if(i>hitbox.ymax)i = hitbox.ymax;
		if(Game::map->getMaterial(coordinate(vec2{newPos.x+(newPos.x < lastPos.x ? hitbox.xmin : hitbox.xmax),newPos.y+i}), true)->solid) {
			return false;
		}
	}
	return true;
}

} /* namespace Pixelverse */
